wdl/imputation.wdl

```diff
@@ -122,6 +150,454 @@ workflow qc_imputation {
+        cpu: 6
+        disks: "local-disk 500 HDD"
+        preemptible: 2
+        zones: "europe-west1-b europe-west1-c europe-west1-d"
+    }
+}
+
+task gather_snpstats_joint_qc {
+
+    String name
+    Array[File] snpstats
+    Array[File] exclude_upstream_variants
+    Array[File] exclude_variants_batch_missingness
+    Float af
+    Float hw
+    Float escape_hwe_maf
+    Float variant_missing_overall
+    Int non_pass_n_batches
+    Int missingness_n_batches
+    String docker
+
+    command <<<
+
+        set -euxo pipefail
+
+        for file in ${sep=" " exclude_upstream_variants}; do
+            cat $file >> ${name}.exclude_upstream_variants.txt
+        done
+
+        # add variants that were non-PASS in the original data in given number of batches
+        grep batch_non_pass ${name}.exclude_upstream_variants.txt | \
+        datamash -s -g 1 count 2 | \
+        awk 'BEGIN{OFS="\t"} $2>=${non_pass_n_batches} {print $1,"joint_qc_batch_non_pass",$2}' \
+        > ${name}.exclude_variants_joint_qc.txt
+
+        # add variants that were PASS in the original data but didn't pass in in original conversion from vcf to plink (practically CNV and some multiallelics)
+        grep -v batch_non_pass ${name}.exclude_upstream_variants.txt | \
+        awk 'BEGIN{OFS="\t"} !seen[$0]++ {$2="joint_qc_"$2; print $0}' \
+        >> ${name}.exclude_variants_joint_qc.txt
+
+        # add variants that have high missingness in given number of batches
+        for file in ${sep=" " exclude_variants_batch_missingness}; do
+            cat $file >> ${name}.exclude_variants_batch_missingness.txt
+        done
+        datamash -s -g 1 count 2 < ${name}.exclude_variants_batch_missingness.txt | \
+        awk 'BEGIN{OFS="\t"} $2>=${missingness_n_batches} {print $1,"joint_qc_batch_high_missing",$2}' \
+        >> ${name}.exclude_variants_joint_qc.txt
+
+        # combine snpstats to one file
+        head -1 ${snpstats[0]} > ${name}.snpstats.txt
+        for file in ${sep=" " snpstats}; do
+            tail -n+2 $file >> ${name}.snpstats.txt
+        done
+
+        # add variants based on snpstats with given thresholds
+        awk '
+        BEGIN {OFS="\t"} NR==1 {for (i=1;i<=NF;i++) h[$i]=i}
+        NR>1 {
+            id=$h["rsid"];
+            af1=$h["alleleA_frequency"];
+            af2=$h["alleleB_frequency"];
+            hw=$h["HW_exact_p_value"];
+            miss=$h["missing_proportion"];
+            maf=$h["minor_allele_frequency"];
+            exp_hom=maf*maf*$h["total"];
+            if (af1<0.5) {
+               hom=$h["AA"]
+            } else {
+               hom=$h["BB"]
+            }
+            if (af1<${af}) print id,"joint_qc_alleleA_frequency",af1;
+            if (af2<${af}) print id,"joint_qc_alleleB_frequency",af2;
+            # allow rare variants with deficiency in homozygotes to escape hw
+            if ((hw+0)<${hw} && (maf > ${escape_hwe_maf} || hom>exp_hom)) print id,"joint_qc_HW_exact_p_value",hw;
```

Comment:
would be good to add comment about the hwe+0 so the purpose is clear why it's needed

Reply:
too easy to be a good janitor and remove "unnecessary code" if outsider makes modifications or amnesia hits